src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,539 @@
+    {
+        // constants
+        private const int ClientNonceLength = 32;
+
+        // static properties
+        /// <summary>
+        /// Gets the name of the mechanism.
+        /// </summary>
+        /// <value>
+        /// The name of the mechanism.
+        /// </value>
+        public static string MechanismName
+        {
+            get { return "MONGODB-AWS"; }
+        }
+
+        // constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MongoAWSAuthenticator"/> class.
+        /// </summary>
+        /// <param name="credential">The credentials.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(UsernamePasswordCredential credential, IEnumerable<KeyValuePair<string, string>> properties)
+            : this(credential, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MongoAWSAuthenticator"/> class.
+        /// </summary>
+        /// <param name="username">The username.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(string username, IEnumerable<KeyValuePair<string, string>> properties)
+            : this(username, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+            : base(CreateMechanism(credential, properties, randomByteGenerator, clock))
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, clock))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGODB-AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
```

Comment:
Refactor this method to:

    private static MongoAWSMechanism CreateMechanism(
        string username,
        SecureString password,
        IEnumerable<KeyValuePair<string, string>> properties,
        IRandomByteGenerator randomByteGenerator,
        IClock clock)
    {
        var awsCredentials =
            CreateAwsCredentialsFromMongoCredentials(username, password, properties) ??
            CreateAwsCredentialsFromEnvironmentVariables() ??
            CreateAwsCredentialsFromEcsResponse() ??
            CreateAwsCredentialsFromEc2Response();

        if (awsCredentials == null)
        {
            throw new InvalidOperationException("Unable to find credentials for MONGODB-AWS authentication.");
        }

        return new MongoAWSMechanism(awsCredentials, randomByteGenerator, clock);
    }